### DISCUSSION TITLE
Column-level privileges feedback

### DISCUSSION THREAD

**Discussion by alaister** [supabase/supabase#20295] [General]
We're excited to announce the release of the column-level privileges management UI: It allows you to grant or revoke privileges on tables and columns based on user roles.

<img src="https://github.com/supabase/supabase/assets/19742402/1c1ff3da-382f-4423-a072-265d7a6dc6b6" width="600" />

What do you like most about it? Least?
Are there any features you find particularly useful?
Anything you think could be improved?
Anything you think is missing?

Happy to hear your thoughts, suggestions, and experiences.

**Comment by roo12312** [comment]
Hi, custom roles are not being shown in the select option

**Comment by markledwich2** [comment]
I can't find this in the product atm

**Reply by GaryAustin1** [reply]
It is a preview feature.   Down on the bottom left is a "person" icon.  Click that and you will see a preview option.

**Comment by FilipeMarch** [comment]
The UI does not work.

When I click on `Apply Changes`, it resets everything to the original state

https://github.com/supabase/supabase/assets/23220309/b17d78fb-9217-472b-8932-13780111536e

Has anyone ever used this feature?

**Reply by GaryAustin1** [reply]
It applies correctly for me.
![image](https://github.com/supabase/supabase/assets/54564956/50d5def2-ded7-443d-a4d6-af120e660a27)

You might check the Postgres logs for errors that are silently stopping the process.  Do you use Prisma (which messes with grants)?

**Reply by FilipeMarch** [reply]
> It applies correctly for me.
> ![image](https://github.com/supabase/supabase/assets/54564956/50d5def2-ded7-443d-a4d6-af120e660a27)
> 
> You might check the Postgres logs for errors that are silently stopping the process.  Do you use Prisma (which messes with grants)?
> 
> 

The error happens when your table is named "users".
I don't use Prisma.

**Comment by joaoguidev** [comment]
Hello.

Great feature! 

The only thing is that my custom roles are not there to be selected.

The only roles available are the usual, anon, service_role and authenticated.

**Comment by ofekd** [comment]
Can you explain why you treat this as an advanced feature and recommend avoiding it if possible? Is that the Alpha state or some inherent thing about it?

**Reply by od-vennre** [reply]
> Can you explain why you treat this as an advanced feature and recommend avoiding it if possible? Is that the Alpha state or some inherent thing about it?

I would like to ask the same question. This is an incredibly important question for those wanting to incorporate this functionality into an application that needs it. Yes, we can build this level of security at the application (e.g. React/Remix) layer but it would make more sense, be more secure, and align better with Supabase and PostgreSQL's philosophies to build it at the DB layer.

Another possible route which has the potential of achieving a similar outcome is using transparently encrypted table columns. Is such a thing supported in Supabase? (I do not think the Vault is the right concept for this, though)

Is discouraging users from using this feature intended to communicate Supabase's lack of desire to evolve it to maturity and hence basically telling us that the only reason we are seeing it is because of a funding-round presentation that required it? or is PostgreSQL's column-level security problematic in Supabase because of some other inherent reason?

For example, does implementing column-specific RLS into the diff tool incorporate a massive workload? or does the `pg_dump` utility not support recreating policies? If there is a technical challenge in getting this awesome feature to work reliably and evolve it to maturity, can you please be transparent about that with the community to help us decide what to do (and not to do) with it?

Thank you.

**Comment by trunghuynh291199** [comment]
How can I hide some columns even when I select (*)?

**Comment by savikko** [comment]
Just as a feedback two "issues" (otherwise seems to be working nicely):
* Lets say we have a table where one user has select permission to subset of the fields. Then, when using Supabase UI "impersonate" feature on Table Editor, if that target user does not have permission to all fields, the UI just times out (even though the request fails with HTTP status 400 - permission denied)
* On Column Level Privileges admin, if i have smaller browser window (vertically), i am unable to select tables which do not fit the dropdown (or actually the window)

**Comment by cerpinn** [comment]
Would this be available on a local Docker instance? I can't find the page nor the preview toggle.

**Reply by mikob** [reply]
The page is here: http://127.0.0.1:54323/project/default/database/column-privileges
But even after enabling it, it doesn't seem to load and CLS doesn't seem to be working correctly for me (select * still gets all the columns)
![image](https://github.com/user-attachments/assets/596b6190-15e1-4f93-a6c8-b0a01217c504)

**Reply by rossbishop** [reply]
Seeing the same as @mikob here - are there any plans to fix this for running locally?

**Reply by alaister** [reply]
Hi,

Thanks for reporting. This should now be fixed, please try updating your CLI.

**Comment by smaldd14** [comment]
I'm having issues with setting CLS using migrations scripts, or via the SQL editor. I've enabled the feature preview, so I have CLS available in my local supabase studio 
<img width="319" alt="Screenshot 2024-12-19 at 9 28 25 PM" src="https://github.com/user-attachments/assets/feee8015-c7b8-4f62-8028-0f513c53c216" />

I have a simple `clubs` table:
```
create table
  public.clubs (
    id uuid not null default gen_random_uuid (),
    slug text not null,
    name character varying(255) not null,
    address text null,
    phone character varying(20) null,
    capacity integer null,
    created_at timestamp with time zone not null default timezone ('utc'::text, now()),
    updated_at timestamp with time zone not null default timezone ('utc'::text, now()),
    image_url text null,
    stripe_customer_id text null,
    auto_pay_enabled boolean null default false,
    auto_pay_max_per_promoter_cents bigint null default 50000,
    constraint clubs_pkey primary key (id),
    constraint clubs_slug_key unique (slug)
  ) tablespace pg_default;
```

And I am trying to restrict access to only allow a certain role to be able to read the `stripe_customer_id` column.

For starters, I had this:
```
-- Revoke SELECT on sensitive columns from public
REVOKE SELECT ON table clubs FROM public;
GRANT SELECT (id, slug, name, address, phone, capacity, image_url, created_at, updated_at) ON table clubs TO public;
```

it returns Success when I run it in SQL editor, but querying `select * from clubs;` still returns all columns no matter what role I impersonate.

I see this warning saying Column privileges are not supported in this version of the Supabase CLI, but I just upgrade to latest version 2.1.1
<img width="997" alt="Screenshot 2024-12-19 at 9 33 12 PM" src="https://github.com/user-attachments/assets/61a02562-3dee-45cc-81c3-6dca12f07148" />


Any ideas on how I can get this to work and test in SQL editor?

**Reply by GaryAustin1** [reply]
The UI should work and SQL to set them should work.
But migrations will not work.  There is no current way around that.
The 3rd party code used for the migrations does not support them.

**Reply by GaryAustin1** [reply]
https://github.com/supabase/cli/issues/1716

**Comment by smaldd14** [comment]
Thanks for the response. SQL editor doesn't seem to do it. I seem to be following the CLS syntax correctly, no?
https://github.com/user-attachments/assets/5517fa94-8aa1-4cc9-bc6a-18715fe57e24
(stripe ID is test ID)

I can confirm that the UI works, but this is not ideal for me

**Reply by GaryAustin1** [reply]
SQL pretty much has to work as it is a postgres thing and not a supabase thing.   Select * should actual error if you have any select grant turned off.  When you use column grants you have to name each row.  You also did not revoke from anon/authenticated just public.  EDIT although I've never tried revoking public for all roles, just using individual roles.

Seems like your table select is still applied though.

**Reply by smaldd14** [reply]
shouldn't `select * ` be returning an error? it instead returns everything like normal. Only when I use the UI and turn off the column, then `select * ` returns an error.

Even doing 
```
-- Revoke SELECT on sensitive columns from public
REVOKE SELECT ON table clubs FROM public;
REVOKE SELECT ON table clubs FROM anon;
REVOKE SELECT ON table clubs FROM authenticated;

GRANT SELECT (id, slug, name, address, phone, capacity, image_url, created_at, updated_at) ON table clubs TO public;
GRANT SELECT (id, slug, name, address, phone, capacity, image_url, created_at, updated_at) ON table clubs TO anon;
GRANT SELECT (id, slug, name, address, phone, capacity, image_url, created_at, updated_at) ON table clubs TO authenticated;
```

doesn't change the result of select *

**Reply by GaryAustin1** [reply]
Yes select * should error if your select on table is revoked and any column is revoked.

Too late for me to test, but this is all postgres and not Supabase so likely something you are doing wrong (or some very new permission bug).    I just don't see it at quick glance.  Seems like the table revoke is not working.  The UI being on or off or at any rev should not impact SQL.

**Reply by GaryAustin1** [reply]
Your code copied exactly for table and last 6 lines with anon user did this for me on latest hosted Supabase.
![image](https://github.com/user-attachments/assets/d70c0f58-2024-48e1-b729-015dce9edfd1)


selecting a single column updated_at did not error.   No data in table.

**Reply by smaldd14** [reply]
Ah I apologize, I had an old RLS policy 
```
CREATE POLICY "Everyone can view basic club information" ON public.clubs
    FOR SELECT USING (true);
```

That I had not deleted locally. If I delete this, now any queries to `select * from clubs;` from anon or authenticated role, just return "success. no rows returned" which makes more sense.

I appreciate the help

**Comment by eqoram** [comment]
I would like to see all roles I created in the dropdown and I think it would be nice to have the option to have a button next to the "Apply Changes" button to get all SQL statements to put these into migration files. Thank you!

EDIT:
Nevermind! I just learned there is no way to assign a custom postgres role to a supabase user. Therefore it doesn't make any sense to display newly created roles in the dropdown. Sadly this also means that you won't be able to assign different kind of column permissions for authenticated users with the Data API (except for the workaround to create different tables).

**Reply by od-vennre** [reply]
Hello, it seems to be possible to assign a custom PostgreSQL role to a Supabase user. Check this link https://supabase.com/docs/guides/storage/schema/custom-roles and also check the `role` column in the `auth.users` Supabase table. It seems as if Supabase would switch roles from authenticator to your custom PostgreSQL role allowing true role-based column-level security. I am curious what you think about this!

**Reply by eqoram** [reply]
@od-vennre Thanks for the heads up! I added the role as mentioned in the linked docs but unfortunately, it still didn't show up Column-level privileges page.

**Comment by mikob** [comment]
BUG: On a local docker instance, this statement doesn't do anything (in the UI or in the db) until I've toggled off at least one of the columns via the UI and saved first.

```
revoke
select
(website, state, country, importance, notes, created_at, last_modified, subscription_id) on table public.orgs
from
authenticated;
```
